client/src/templates/Challenges/generic/show.tsx

```diff
@@ -0,0 +1,274 @@
+  bindActionCreators(
+    {
+      initTests,
+      updateChallengeMeta,
+      challengeMounted,
+      updateSolutionFormValues,
+      openCompletionModal: () => openModal('completion')
+    },
+    dispatch
+  );
+
+// Types
+interface ShowQuizProps {
+  challengeMounted: (arg0: string) => void;
+  data: { challengeNode: ChallengeNode };
+  description: string;
+  initTests: (xs: Test[]) => void;
+  isChallengeCompleted: boolean;
+  openCompletionModal: () => void;
+  pageContext: {
+    challengeMeta: ChallengeMeta;
+  };
+  updateChallengeMeta: (arg0: ChallengeMeta) => void;
+  updateSolutionFormValues: () => void;
+}
+
+const ShowGeneric = ({
+  challengeMounted,
+  data: {
+    challengeNode: {
+      challenge: {
+        assignments,
+        bilibiliIds,
+        block,
+        blockType,
+        description,
+        challengeType,
+        fields: { tests },
+        helpCategory,
+        instructions,
+        title,
+        translationPending,
+        superBlock,
+        videoId,
+        videoLocaleIds
+      }
+    }
+  },
+  pageContext: { challengeMeta },
+  initTests,
+  updateChallengeMeta,
+  openCompletionModal,
+  isChallengeCompleted
+}: ShowQuizProps) => {
+  const { t } = useTranslation();
+  const { nextChallengePath, prevChallengePath } = challengeMeta;
+  const container = useRef<HTMLElement | null>(null);
+
+  const blockNameTitle = `${t(
+    `intro:${superBlock}.blocks.${block}.title`
+  )} - ${title}`;
+
+  useEffect(() => {
+    initTests(tests);
+    updateChallengeMeta({
+      ...challengeMeta,
+      title,
+      challengeType,
+      helpCategory
+    });
+    challengeMounted(challengeMeta.id);
+    container.current?.focus();
+    // This effect should be run once on mount
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  useEffect(() => {
+    updateChallengeMeta({
+      ...challengeMeta,
+      title,
+      challengeType,
+      helpCategory
+    });
+    challengeMounted(challengeMeta.id);
+  }, [
+    title,
+    challengeMeta,
+    challengeType,
+    helpCategory,
+    challengeMounted,
+    updateChallengeMeta
+  ]);
+
+  // video
+  const [videoIsLoaded, setVideoIsLoaded] = useState(false);
+
+  const handleVideoIsLoaded = () => {
+    setVideoIsLoaded(true);
+  };
+
+  // assignments
+  const [allAssignmentsCompleted, setAllAssignmentsCompleted] = useState(false);
+  const [assignmentsCompleted, setAssignmentsCompleted] = useState(0);
+
+  const handleAssignmentChange = (
+    event: React.ChangeEvent<HTMLInputElement>,
+    totalAssignments: number
+  ) => {
+    const newAssignmentsCompleted = event.target.checked
+      ? assignmentsCompleted + 1
+      : assignmentsCompleted - 1;
+
+    setAssignmentsCompleted(newAssignmentsCompleted);
+    setAllAssignmentsCompleted(totalAssignments === newAssignmentsCompleted);
+  };
+
+  // submit
+  const handleSubmit = () => {
+    if (assignments.length > 0 && allAssignmentsCompleted) {
```

Comment:
Are assignments optional? If so:
```suggestion
    if (assignments.length == 0 || allAssignmentsCompleted) {
```
otherwise it's impossible to open the completion modal when there are no assignments.

Reply:
✅